src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java

```diff
@@ -0,0 +1,380 @@
+import org.springframework.beans.factory.annotation.Configurable;
+
+import com.krishagni.catissueplus.core.administrative.domain.ScheduledJobRun;
+import com.krishagni.catissueplus.core.administrative.events.ListPvCriteria;
+import com.krishagni.catissueplus.core.administrative.events.PvDetail;
+import com.krishagni.catissueplus.core.administrative.services.ScheduledTask;
+import com.krishagni.catissueplus.core.biospecimen.domain.CollectionProtocol;
+import com.krishagni.catissueplus.core.biospecimen.domain.CollectionProtocolEvent;
+import com.krishagni.catissueplus.core.biospecimen.events.CollectionProtocolDetail;
+import com.krishagni.catissueplus.core.biospecimen.events.CollectionProtocolEventDetail;
+import com.krishagni.catissueplus.core.biospecimen.events.CollectionProtocolSiteDetail;
+import com.krishagni.catissueplus.core.biospecimen.events.SpecimenRequirementDetail;
+import com.krishagni.catissueplus.core.biospecimen.repository.DaoFactory;
+import com.krishagni.catissueplus.core.biospecimen.services.CollectionProtocolService;
+import com.krishagni.catissueplus.core.common.PlusTransactional;
+import com.krishagni.catissueplus.core.common.errors.OpenSpecimenException;
+import com.krishagni.catissueplus.core.common.events.RequestEvent;
+import com.krishagni.catissueplus.core.common.events.ResponseEvent;
+import com.krishagni.catissueplus.core.common.events.UserSummary;
+import com.krishagni.catissueplus.core.common.service.PermissibleValueService;
+
+@Configurable
+public class CarsImporter implements ScheduledTask {
+	private final static Log logger = LogFactory.getLog(CarsImporter.class);
+	
+	@Autowired
+	private CollectionProtocolService cpSvc;
+	
+	@Autowired
+	private PermissibleValueService pvSvc;
+	
+	@Autowired
+	private DaoFactory daoFactory;
+	
+	private final static String SPECIMEN_STORAGE_TYPE_VIRTUAL = "Virtual";
+	
+	private final static String NOT_SPECIFIED = "Not Specified";
+	
+	@Override
+	public void doJob(ScheduledJobRun jobRun) throws Exception {
+		CarsReader carsReader = new CarsReader();
+		try {
+			while (carsReader.hasDistinctCp()) {
+				importRecords(carsReader);
+			}
+		} catch (Exception e) {
+			logger.error(e);
+		} finally {
+			carsReader.close();
+		}
+	}
+	
+	private void importRecords(CarsReader carsReader) throws Exception {
+		try {
+			CarsCpDetail carsCpDetail = getCarsCpDetail(carsReader);
+			processCpDiff(carsCpDetail);
+		} catch (OpenSpecimenException ose) {
+			logger.error(ose);
+		}
+	}
+	
+	private CarsCpDetail getCarsCpDetail(CarsReader carsReader) {
+		CarsCpDetail carsCpDetail = new CarsCpDetail();
+		
+		while (carsReader.hasCp()) {
+			CarsDetail carsDetail = carsReader.nextCp();
+			
+			CollectionProtocolDetail cp = toCp(carsDetail);
+			CollectionProtocolEventDetail event = toEvent(carsDetail);
+			SpecimenRequirementDetail sr = toSr(carsDetail);
+			
+			carsCpDetail = carsCpDetail.from(cp, event, sr);
+		}
+		return carsCpDetail;
```

Comment:
I don't like this loop structure ...

Something better is like this:

```
while (carsReader.hasNext()) {
  // detail : { cp attrs, events: [ { event attrs, specimenRequirements: [ { req attrs }, ... ]}, .. ]}
  CollectionProtocolDetail input = carsReader.next();
  processCp(input);
}
```

If you feel the reader should not be aware of transforming the input records to CollectionProtocolDetail then the next best option is:

```
while (carsReader.hasNext()) {
  List<CarsRecord> cpRows = carsReader.next();
  CollectionProtocolDetail input = toCpDetail(cpRows);
  processCp(input);
}
```